modules/reindex/src/main/java/org/opensearch/index/reindex/Reindexer.java

```diff
@@ -91,40 +94,76 @@
     private final ScriptService scriptService;
     private final ReindexSslConfig reindexSslConfig;
+    private final Optional<RemoteReindexExtension> remoteExtension;
 
     Reindexer(ClusterService clusterService, Client client, ThreadPool threadPool, ScriptService scriptService,
               ReindexSslConfig reindexSslConfig) {
+        this(clusterService, client, threadPool, scriptService, reindexSslConfig, Optional.empty());
+    }
+
+    Reindexer(ClusterService clusterService, Client client, ThreadPool threadPool, ScriptService scriptService,
+              ReindexSslConfig reindexSslConfig, Optional<RemoteReindexExtension> remoteExtension) {
         this.clusterService = clusterService;
         this.client = client;
         this.threadPool = threadPool;
         this.scriptService = scriptService;
         this.reindexSslConfig = reindexSslConfig;
+        this.remoteExtension = remoteExtension;
     }
 
     public void initTask(BulkByScrollTask task, ReindexRequest request, ActionListener<Void> listener) {
         BulkByScrollParallelizationHelper.initTaskState(task, request, client, listener);
     }
 
     public void execute(BulkByScrollTask task, ReindexRequest request, ActionListener<BulkByScrollResponse> listener) {
+        ActionListener<BulkByScrollResponse> remoteReindexActionListener = getRemoteReindexWrapperListener(listener, request);
         BulkByScrollParallelizationHelper.executeSlicedAction(task, request, ReindexAction.INSTANCE, listener, client,
             clusterService.localNode(),
             () -> {
                 ParentTaskAssigningClient assigningClient = new ParentTaskAssigningClient(client, clusterService.localNode(), task);
                 AsyncIndexBySearchAction searchAction = new AsyncIndexBySearchAction(task, logger, assigningClient, threadPool,
-                    scriptService, reindexSslConfig, request, listener);
+                    scriptService, reindexSslConfig, request, remoteReindexActionListener, getInterceptor(request));
                 searchAction.start();
             });
 
     }
 
+    private Optional<HttpRequestInterceptor> getInterceptor(ReindexRequest request) {
+        if (request.getRemoteInfo() == null) {
+            return Optional.empty();
+        } else {
+            return remoteExtension.map(x -> x.getInterceptorProvider()).flatMap(provider ->
+                provider.getRestInterceptor(request, threadPool.getThreadContext()));
+        }
+    }
+
+    private ActionListener<BulkByScrollResponse> getRemoteReindexWrapperListener(
+        ActionListener<BulkByScrollResponse> listener, ReindexRequest reindexRequest) {
+        if (reindexRequest.getRemoteInfo() == null) {
+            return listener;
+        }
+        if (remoteExtension.isPresent()) {
+            return remoteExtension.get().getRemoteReindexActionListener(listener, reindexRequest);
+        }
+        logger.error("Not able to load remote reindex metrics listener");
```

Comment:
This error message seems confusing, in case there are no actual extensions present to the Reindexer plugin.

Reply:
Changed the message